Generalize and improve docstrings of node_link.py

Michael-E-Rose opened the pull request:
This is related to issue #1951.

Changes: 
* Set default value of `id` to `name` and rename attribute key `id` to `name`
* Use `attrs` parameter to update `_attrs` such that default parameters don't need to be provided
* Add attribute key `link` with default value `links`
* Update docstrings (i.e. extend example, apply PEP8 and remove obsolete note)
* Apply PEP8 to `test_node_link.py`

Some questions remain:
1. Do we really need `make_str`? After all, it's just a wrapper and doesn't make things shorter. Removing it means to import one package less.
2. Should we get rid of all the blanks in the output? Especially for large graphs writing `{'source':0,'target':1}` instead of `{'source': 0, 'target': 1}` results in a good compression.
3. Should I remove `__author__` attribute according to #1728?

dschult commented:
1) Look at the ```make_str``` code. It is supposed to help with python2/python3 unicode/str handling.  It's goal is not to make code shorter. It is to centralize and simplify handling of strings/unicode/etc.

2) We don't strive for compression of code. We strive for readability of code. Follow PEP8 guidelines (including the idea that exceptions to the guidelines are sometimes better than the guidelines).  We want readable, maintainable code.

3) Author attributes are NOT being removed. They are being replaced by comment lines. See the section of the NetworkX GitHub Wiki page for how to replace the ```__author__``` code with ```# Authors:``` comments

Thanks

Michael-E-Rose commented:
ad 2) I'm speaking of the output, not this code. For a graph of 500 nodes, no one is looking into the resulting .json file. But the sheer amount of blanks heavily increases the weight of the file.

ad 3) That's exactly what I mean. Hence my link to the issue.

I'll update it just now and fix the tests/checks.

Michael-E-Rose commented:
Can someone give a hint why all checks fail? A networkx error is raised because the keys `source`, `target` and `key` are not unique, but I can't see what that should be the case. I didn't even touch these entries and can import the updated method without error.

jfinkels commented on networkx/readwrite/json_graph/node_link.py:
```diff
@@ -1,18 +1,21 @@
+              key='key', link='links')
 
 
 def node_link_data(G, attrs=_attrs):
```
Using a mutable default value for a keyword argument will cause problems. Use `attrs=None` instead, and have a special check at the beginning of the function for `if attrs is None: ...`.

Michael-E-Rose replied:
Alright, thanks. But how can this cause problems *now*? `attrs=_attrs` was
there before.
Am 22.02.2016 22:12 schrieb "jfinkels" <notifications@github.com>:

> In networkx/readwrite/json_graph/node_link.py
> <https://github.com/networkx/networkx/pull/2000#discussion_r53693600>:
>
> >  __all__ = ['node_link_data', 'node_link_graph']
> >
> >
> > -_attrs = dict(id='id', source='source', target='target', key='key')
> > +_attrs = dict(source='source', target='target', name='name',
> > +              key='key', link='links')
> >
> >
> >  def node_link_data(G, attrs=_attrs):
>
> Using a mutable default value for a keyword argument will cause problems.
> Use attrs=None instead, and have a special check at the beginning of the
> function for if attrs is None: ....
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/2000/files#r53693600>.
>

jfinkels replied:
You added a `_attrs.update()` below.

Michael-E-Rose commented:
Now a problem remains with `node_link_data`'s docstring. I added the content of `json_graph.node_link_data()`. On python 3.x it is receiving a different order than the one I added. Seeing the result of some networkx commands in examples always helped me but if it's causing errors, maybe I should reduce the example in the dostring?

jfinkels commented:
You could evaluate each pair individually in the docstring:

    >>> data['directed']
    False
    >>> data['graph']
    {}
    ...

Michael-E-Rose commented:
So I removed the printout from the docstring examples, although they would be helpful. It's because the printouts are dictionaries and Travis CI sometimes gets a different order. I reported a bug with travis-ci.

jfinkels commented:
That's not really a bug, dictionaries are unordered in Python by design. Search "Python dictionary doctest" for some discussion/solutions.

Michael-E-Rose commented:
I meant, the bug is that Travis CI complains about that.

hagberg commented:
No - we need to handle that with our tests, not Travis CI.

On Tue, Feb 23, 2016 at 9:17 AM, Michael E. Rose <notifications@github.com>
wrote:

> I meant, the bug is that Travis CI complains about that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/networkx/networkx/pull/2000#issuecomment-187769593>.
>

Michael-E-Rose commented:
I could write `sorted(data1.items())` or leave it like this (that is, without a printout).

dschult commented:
```sorted(data1.items())``` seems good to me.
Also, the Travis-CI tests should give the same results as tests using ```nosetests``` on your own machine. So it really isn't a Travis-CI issue. 

You can install ```nose``` and then run ```nosetests -v --with-doctest``` to check your work before committing. (you need to use environment variable PYTHONPATH or similar to make sure python will find your networkx directory)

Michael-E-Rose commented:
I gave up and removed the example output from the docstring. Reason: Travis' behavior is completely unpredictable: Sometimes tests pass except when using miniconda, and sometimes tests pass on Python 2.7 only. Since I'm on Python 2.7 as well, `nosetests -v --with-doctest` didn't help me in predicting what Travis CI would find with Python 3.3.

hagberg commented:
It's not unpredictability in the Travis testing causing your test failures.  It is the unpredictable output ordering of Python dictionaries using Python 3.  We are obligated to write our tests to handle that.  I suggest to start using Python 3 when you can.
